app/javascript/mastodon/features/interaction_modal/index.jsx

```diff
@@ -18,62 +25,207 @@ const mapDispatchToProps = (dispatch) => ({
   },
 });
 
-class Copypaste extends React.PureComponent {
+const PERSISTENCE_KEY = 'mastodon_home';
+
+const isValidDomain = value => {
+  const url = new URL('https:///path');
+  url.hostname = value;
+  return url.hostname === value;
+};
+
+const valueToDomain = value => {
+  // If the user uses a URL to their profile page or server
+  if (/^https?:\/\//.test(value)) {
+    try {
+      return (new URL(value)).host;
+    } catch {
+      return '';
+    }
+  // If the user writes their full handle including username
+  } else if (value.includes('@')) {
+    const segments = value.split('@');
+
+    if (segments.length === 2) {
+      return segments[1];
+    } else {
+      return '';
+    }
+  }
+
+  return value;
+};
+
+class LoginForm extends React.PureComponent {
 
   static propTypes = {
-    value: PropTypes.string,
+    resourceUrl: PropTypes.string,
+    intl: PropTypes.object.isRequired,
   };
 
   state = {
-    copied: false,
+    value: localStorage ? (localStorage.getItem(PERSISTENCE_KEY) || '') : '',
+    expanded: false,
+    selectedOption: -1,
+    isLoading: false,
+    error: false,
+    options: [],
   };
 
   setRef = c => {
     this.input = c;
   };
 
-  handleInputClick = () => {
-    this.setState({ copied: false });
-    this.input.focus();
-    this.input.select();
-    this.input.setSelectionRange(0, this.input.value.length);
+  handleChange = ({ target }) => {
+    this.setState({ value: target.value, isLoading: true, error: false }, () => this._loadOptions());
+  };
+
+  handleSubmit = () => {
+    const { value } = this.state;
+    const { resourceUrl } = this.props;
+
+    const domain = valueToDomain(value);
+
+    if (!isValidDomain(domain)) {
+      this.setState({ error: true });
+      return;
+    }
+
+    if (localStorage) {
+      localStorage.setItem(PERSISTENCE_KEY, domain);
+    }
+
+    const redirectUrl = new URL(`https://${domain}/authorize_interaction?uri=${encodeURIComponent(resourceUrl)}`);
```

Comment:
This will only work with Mastodon, not Pleroma, Akkoma, Misskey, Calckey, or PixelFed, while some of these do support the old flow.

People using any of these services will have either a 404, or in some case, end up on the profile of an account called `authorize_interaction` if it exists.

Reply:
If I use a new .well-known URL now, it won't work for most people on release.

Reply:
That's true… that and unambiguous support for configurations where `LOCAL_DOMAIN` ≠ `WEB_DOMAIN` are the reasons why I think we should rather ask for the handle rather than the server, this would allow us to use the already well-supported flow.

Reply:
It worked. And protocol-wise, that's not “just to get the second part”. Multiple AP software can theoretically run on the same domain. And more long-term thinking, if we ever move to DIDs or even just to have multi-tenancy and e.g. mail providers allowing you to point to your Mastodon instance to have the same fediverse handle as your mail address, we can't rely solely on the domain.

Reply:
I think it's much easier to type a domain with autocomplete suggestions than to type your entire username (my instinct is that the autocomplete will play a major role in how user-friendly this feature will be perceived), I also don't think it's necessarily good to require people to give away their identity for this. Most apps require only the domain for the login. I would love to deploy this and observe how people get along with it in practice.

Reply:
could we do this via webfinger against the bare domain name as resolved via the webfinger host? i know conceptually it is only 100% unambiguous if given either an acct: or an actor id, because then you could define something like what i described above with proxyUrl. but if the important bit is specifically "put in a domain name only" then you could still get there by having the domain resolve via either webfinger or activitypub. (claire raises a good point about the differing domains and also different softwares on the same domain, which i'm sure cause problems in other parts of the code as well, but one thing at a time)

Reply:
> I think it's much easier to type a domain with autocomplete suggestions than to type your entire username (my instinct is that the autocomplete will play a major role in how user-friendly this feature will be perceived)

I'm not completely sure. And I think autocomplete on handle could even be more useful, but it's much less likely to already know about your account, so it's definitely not a clear win either.

That being said, we can support both: if it looks like a handle, autocomplete on account search, otherwise, autocomplete on domain. You could even imagine autocompleting the domain part if no user matches.

> I also don't think it's necessarily good to require people to give away their identity for this

That is a very good point, although the modal's point is for you to interact with the resource from your remote account, so in most cases they'd give away their identity on the next step.

> Most apps require only the domain for the login.

True, and it can already cause quite some confusion when `WEB_DOMAIN` ≠ `LOCAL_DOMAIN`.

> I would love to deploy this and observe how people get along with it in practice.

I agree, but I would also like not committing to something that is neither future-proof nor compatible with the non-Mastodon fediverse.

> could we do this via webfinger against the bare domain name as resolved via the webfinger host?

I don't think that's currently supported, so that would present most of the same drawbacks as the current PR while also breaking compatibility with current Mastodon :/